src/AstTypeAnalysis.cpp

```diff
@@ -476,251 +432,244 @@ AstClause* createAnnotatedClause(
-                    case AstNumericConstant::Type::Int:
-                        if (canBeParsedAsRamSigned(constant.getConstant())) {
-                            possibleTypes.insert(env.getNumberType());
-                        }
-                        break;
-                    case AstNumericConstant::Type::Uint:
-                        if (canBeParsedAsRamUnsigned(constant.getConstant())) {
-                            possibleTypes.insert(env.getUnsignedType());
-                        }
-                        break;
-                    case AstNumericConstant::Type::Float:
-                        if (canBeParsedAsRamFloat(constant.getConstant())) {
-                            possibleTypes.insert(env.getFloatType());
-                        }
-                        break;
-                }
-                // Else: all numeric types that can be parsed are valid.
-            } else {
-                if (canBeParsedAsRamSigned(constant.getConstant())) {
-                    possibleTypes.insert(env.getNumberType());
-                }
+    void visitBinaryConstraint(const AstBinaryConstraint& rel) override {
+        auto lhs = getVar(rel.getLHS());
+        auto rhs = getVar(rel.getRHS());
+        addConstraint(isSubtypeOf(lhs, rhs));
+        addConstraint(isSubtypeOf(rhs, lhs));
+    }
 
-                if (canBeParsedAsRamUnsigned(constant.getConstant())) {
-                    possibleTypes.insert(env.getUnsignedType());
+    void visitFunctor(const AstFunctor& fun) override {
+        auto functorVar = getVar(fun);
+
+        // In polymorphic case
+        // We only require arguments to share a base type with a return type.
+        // (instead of, for example, requiring them to be of the same type)
+        // This approach is related to old type semantics
+        // See #1296 and tests/semantic/type_system4
+        if (auto intrinsicFunctor = dynamic_cast<const AstIntrinsicFunctor*>(&fun)) {
+            if (isOverloadedFunctor(intrinsicFunctor->getFunction())) {
+                for (auto* argument : intrinsicFunctor->getArguments()) {
+                    auto argumentVar = getVar(argument);
+                    addConstraint(subtypesOfTheSameBaseType(argumentVar, functorVar));
                 }
 
-                if (canBeParsedAsRamFloat(constant.getConstant())) {
-                    possibleTypes.insert(env.getFloatType());
-                }
+                return;
             }
-
-            addConstraint(hasSuperTypeInSet(getVar(constant), possibleTypes));
         }
 
-        // binary constraint
-        void visitBinaryConstraint(const AstBinaryConstraint& rel) override {
-            auto lhs = getVar(rel.getLHS());
-            auto rhs = getVar(rel.getRHS());
-            addConstraint(isSubtypeOf(lhs, rhs));
-            addConstraint(isSubtypeOf(rhs, lhs));
+        // Debug report does not run the standard transformers.
```

Comment:
You could simplify this to just say "The type of the user-defined ...". The analysis is independent of the debug report.